Elements/src/Spatial/AdaptiveGrid/AdaptiveGrid.cs

```diff
@@ -1330,14 +1330,110 @@ private void SplitGrid(Grid2d grid, IEnumerable<Vector3> keyPoints)
+            var vDomain = grid.V.curveDomain;
+            Vector3 xAxis = (grid.U.Evaluate(uDomain.Max) - grid.U.Evaluate(uDomain.Min)) / uDomain.Length;
+            Vector3 yAxis = (grid.V.Evaluate(vDomain.Max) - grid.V.Evaluate(vDomain.Min)) / vDomain.Length;
+            Vector3 zAxis = xAxis.Cross(yAxis);
+            var fromGrid = new Transform(grid.U.Evaluate(uDomain.Min) + grid.V.Evaluate(vDomain.Min), xAxis, yAxis, zAxis).Concatenated(grid.fromGrid);
+            return fromGrid;
+        }
+
+        private List<(Vector3 from, Vector3 to)> splitSegmentWithPoints((Vector3 from, Vector3 to) line, List<Vector3> points)
+        {
+            var resultingSegments = new List<(Vector3 from, Vector3 to)>();
+            var lineVector = line.to - line.from;
+            points.Add(line.from);
+            points.Add(line.to);
+            points.Sort((p, q) => Math.Sign(lineVector.Dot(p - q)));
+            for (int i = 1; i < points.Count; ++i)
+            {
+                resultingSegments.Add((points[i - 1], points[i]));
+            }
+            return resultingSegments;
+        }
+
+        private List<(Vector3 from, Vector3 to)> splitSegmentsWithPoints(List<Line> segments, double u, List<double> coords, bool coordsAreX, List<Vector3> intersectionPoints)
+        {
+            var swapXYAxes = new Transform(new Vector3(0, 0, 0), Vector3.YAxis, Vector3.XAxis, Vector3.ZAxis);
+            if (coordsAreX)
+            {
+                segments = segments.Select(l => l.TransformedLine(swapXYAxes)).ToList();
+            }
+
+            segments = segments.Select(l => l.Start.Y < l.End.Y ? l : l.Reversed()).ToList();
+            segments.Sort((l1, l2) => l1.Start.Y.CompareTo(l2.Start.Y));
+            var resultingSegments = new List<(Vector3 from, Vector3 to)>();
+            var newIntersectionPoints = new List<Vector3>();
+
+            for (int segmentId = 0, yId = 0; segmentId < segments.Count; ++segmentId)
+            {
+                var previousPoint = segments[segmentId].Start;
+                var lastPoint = segments[segmentId].End;
+                newIntersectionPoints.Add(previousPoint);
+                newIntersectionPoints.Add(lastPoint);
+                while (yId < coords.Count && coords[yId] < previousPoint.Y)
+                {
+                    ++yId;
+                }
+                while (yId < coords.Count && coords[yId] < lastPoint.Y)
+                {
+                    var nextPoint = new Vector3(u, coords[yId]);
+                    newIntersectionPoints.Add(nextPoint);
+                    resultingSegments.Add((previousPoint, nextPoint));
+                    previousPoint = nextPoint;
+                    ++yId;
+                }
+                resultingSegments.Add((previousPoint, lastPoint));
+            }
+
+            if (coordsAreX)
+            {
+                segments = segments.Select(l => l.TransformedLine(swapXYAxes)).ToList();
```

Comment:
I don't think we need to transform them back. segments is already different list than was passed into the function originally.